lib/reversetunnel/localsite.go

```diff
@@ -446,30 +450,40 @@ func (s *localSite) handleHeartbeat(rconn *remoteConn, ch ssh.Channel, reqC <-ch
 }
 
 func (s *localSite) getRemoteConn(dreq *sshutils.DialReq) (*remoteConn, error) {
-	s.Lock()
-	defer s.Unlock()
+	s.remoteConnsMtx.Lock()
+	defer s.remoteConnsMtx.Unlock()
 
 	// Loop over all connections and remove and invalid connections from the
 	// connection map.
-	for key := range s.remoteConns {
-		if s.remoteConns[key].isInvalid() {
+	for key, conns := range s.remoteConns {
+		validConns := conns[:0]
```

Comment:
Is this generally preferred to any other method of obtaining an empty slice?

Reply:
No, this is very much a specific trick to filter items out of a slice in-place if you know that you're the only reference to the underlying buffer. `validConns` is an empty slice at the beginning of `conns`, so appending to it will overwrite the same buffer instead of allocating a new one. This would blow up in a very weird way if we ever returned a slice to the same memory, but here we're the only ones in possession of those buffers and we're only returning copies of the values, so we're good - and avoiding allocations here it's kind of important, as this happens for all existing tunnels on every `getRemoteConn`.